lib/HLSL/HLOperationLower.cpp

```diff
@@ -6856,34 +6870,32 @@ void GenerateStructBufSt(Value *handle, Value *bufIdx, Value *offset,
   Builder.CreateCall(dxilF, Args);
 }
 
-Value *TranslateStructBufMatLd(Type *matType, IRBuilder<> &Builder,
-                               Value *handle, hlsl::OP *OP, Value *status,
-                               Value *bufIdx, Value *baseOffset,
-                               const DataLayout &DL) {
-  HLMatrixType MatTy = HLMatrixType::cast(matType);
-  Type *EltTy = MatTy.getElementTypeForMem();
-  unsigned  EltSize = DL.getTypeAllocSize(EltTy);
+
+static Value* TranslateStructBufVecLd(Type* VecEltTy, unsigned ElemCount,
+  IRBuilder<>& Builder, Value* handle, hlsl::OP* OP, Value* status,
+  Value* bufIdx, Value* baseOffset, const DataLayout& DL, std::vector<Value*> &bufLds, bool isScalarTy) {
+  unsigned  EltSize = DL.getTypeAllocSize(VecEltTy);
   Constant* alignment = OP->GetI32Const(EltSize);
 
-  Value *offset = baseOffset;
+  Value* offset = baseOffset;
   if (baseOffset == nullptr)
     offset = OP->GetU32Const(0);
 
-  unsigned matSize = MatTy.getNumElements();
-  std::vector<Value *> elts(matSize);
-
-  unsigned rest = (matSize % 4);
+  std::vector<Value*> elts(ElemCount);
+  unsigned rest = (ElemCount % 4);
   if (rest) {
-    Value *ResultElts[4];
-    GenerateStructBufLd(handle, bufIdx, offset, status, EltTy, ResultElts, OP, Builder, 3, alignment);
+    Value* ResultElts[4];
+    Value *bufLd = GenerateStructBufLd(handle, bufIdx, offset, status, VecEltTy, ResultElts, OP, Builder, rest, alignment);
```

Comment:
Just realized, and this was already present in the refactored code:
While the code produces correct results, when rest is non-zero, it will produce 4-vector loads that aren't aligned in an expected way to the beginning of the vector/matrix, leading to possible performance implications.  We should fix this in a subsequent change, since It may be a pattern used elsewhere as well.

Reply:
Given that the noted codegen behavior existed before this change, I have another work item to track this #3457 so that we can introduce this as a separate change.